File: neetcode-250/sliding-window/219-contains-duplicate-II.ts
function containsNearbyDuplicate(nums: number[], k: number): boolean {
	const seen = new Set([nums[0]]);
	let i = 0;
	for (let j = 1; j < nums.length; ++j) {
		if (j - i > k) {
			seen.delete(nums[i]);
			++i;
		}

		if (seen.has(nums[j])) {
			return true;
		}

		seen.add(nums[j]);
	}

	return false;
}

File: neetcode-250/sliding-window/219-contains-duplicate-II.py
from typing import List


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        seen, left = set([nums[0]]), 0

        for right in range(1, len(nums)):
            left_n, right_n = nums[left], nums[right]

            if right - left > k:
                seen.remove(left_n)
            left += 1

            if right_n in seen:
                return True

            seen.add(right_n)

        return False
